src/oxli/storage.cc

```diff
@@ -95,6 +95,33 @@ void BitStorage::update_from(const BitStorage& other)
     }
 }
 
+double BitStorage::similarity(const BitStorage& other)
+{
+    if (_tablesizes != other._tablesizes) {
+        throw oxli_exception("both nodegraphs must have same table sizes");
+    }
+
+    uint64_t intersection = 0;
+    uint64_t union_size = 0;
+    for (unsigned int table_num = 0; table_num < _n_tables; table_num++) {
+        Byte * me = _counts[table_num];
+        Byte * ot = other._counts[table_num];
+        uint64_t tablesize = _tablesizes[table_num];
+        uint64_t tablebytes = tablesize / 8 + 1;
+
+        for (uint64_t index = 0; index < tablebytes; index++) {
+            // First, get how many values in common we have
+            intersection += __builtin_popcountll(me[index] & ot[index]);
+            union_size += __builtin_popcountll(me[index] | ot[index]);
```

Comment:
Let me see if I'm grokking this:

- bitwise AND --> buckets where both Bloom filters have 1
- bitwise OR --> buckets where either Bloom filter has 1
- `__builtin_popcountll` --> optimized function for counting buckets that satisfy the specified criteria

Reply:
`__builtin_popcountll` is a glorified count function to check how many bits are set to 1. A pure python equivalent would be
```
>>> a = 0b0101
>>> b = 0b1010
>>> bin(a & b).count('1')
0
>>> bin(a | b).count('1')
4
```